File: Leetcode 2801-2900/2872/2872_1.py
# Leetcode 2872: Maximum Number of K-Divisible Components
# https://leetcode.com/problems/maximum-number-of-k-divisible-components/
# Solved on 21st of December, 2024

class Solution:
    def dfs(self, graph: list[list[int]], u: int, prev: int, values: list[int], k: int) -> int:
        tree_sum = values[u]

        for v in graph[u]:
            if v != prev:
                tree_sum += self.dfs(graph, v, u, values, k)

        if tree_sum % k == 0:
            self.ans += 1

        return tree_sum

    # Main Function
    def maxKDivisibleComponents(self, n: int, edges: list[list[int]], values: list[int], k: int) -> int:
        """
        Given a tree with n nodes where each node has a value, and an integer k,
        we need to find the maximum number of k-divisible components we can obtain
        by removing some edges. A k-divisible component is a connected component
        where the sum of its node values is divisible by k.

        Args:
            n: The number of nodes in the tree.
            edges: A list of edges representing the tree.
            values: A list of values for each node.
            k: The integer to check for divisibility.
        """
        graph = [[] for _ in range(n)]

        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        self.ans = 0
        self.dfs(graph, 0, -1, values, k)
        return self.ans

File: Leetcode 2801-2900/2872/2872-1.java
// Leetcode 2872: Maximum Number of K-Divisible Components
// https://leetcode.com/problems/maximum-number-of-k-divisible-components/
// Solved on 28th of November, 2025
import java.util.ArrayList;
import java.util.List;

class Solution {
    private int count;

    /**
     * Calculates the maximum number of k-divisible components in a tree.
     * A component is k-divisible if the sum of values of its nodes is divisible by k.
     *
     * @param n The number of nodes in the tree.
     * @param edges An array of edges representing the tree structure.
     * @param values An array where values[i] is the value of the i-th node.
     * @param k The divisor.
     * @return The maximum number of k-divisible components.
     */
    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        
        count = 0;
        dfs(0, -1, adj, values, k);
        return count;
    }

    private long dfs(int node, int parent, List<List<Integer>> adj, int[] values, int k) {
        long sum = values[node];
        for (int neighbor : adj.get(node)) {
            if (neighbor != parent) {
                sum += dfs(neighbor, node, adj, values, k);
            }
        }
        
        if (sum % k == 0) {
            count++;
            return 0;
        }
        return sum;
    }
}